### PR TITLE
Drops compile dep on archiver/v3

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This removes the dependency on "archiver/v3" from the binary because
* It has no support for streaming https://github.com/mholt/archiver/pull/199, so we had to buffer to disk first
* Including "archiver/v3" makes it easy to accidentally bloat the binary (we did this by 25% before)
* The result is simpler dependencies

Before:
```bash
$ du -sk build/bin/*/*/getenvoy
8020	build/bin/darwin/amd64
7528	build/bin/linux/amd64
```

```bash
$ du -sk build/bin/*/*/getenvoy
7728	build/bin/darwin/amd64
7264	build/bin/linux/amd64
```

### CODE REVIEW

File: pkg/internal/targz.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 Tetrate
```
Comment (codefromthecrypt) [inline pkg/internal/targz.go:1]:
```suggestion
// Copyright 2021 Tetrate
```

File: pkg/internal/targz.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"archive/tar"
+	"compress/gzip"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// TarGZ makes archives the source as the dstination filename. This avoids a large (~3MB) dependency on archiver/v3.
+func TarGZ(dst, src string) error { // dst, src order like io.Copy
```
Comment (codefromthecrypt) [inline pkg/internal/targz.go:27]: note I'm using implicit test coverage for now, and when I looked at it, the places not touched will be hard, if possible to reproduce (ex tar err when the input is a unix pipe). However, I'm fine making unit tests for this anyway if we want to even if it doesn't improve statement coverage.

File: pkg/test/manifest/server.go [hunk 3]
```diff
@@ -26,19 +26,19 @@ import (
...
 
 const archiveFormat = ".tar.gz"
+const buildsPath = "/builds/"
 
 // RequireManifestTestServer serves "/manifest.json", which contains download links a compressed archive of artifactDir
 func RequireManifestTestServer(t *testing.T, m *api.Manifest) *httptest.Server {
 	s := &server{t: t}
 	h := httptest.NewServer(s)
-	s.manifest = rewriteDownloadLocations(t, h.URL, m)
+	s.rewriteDownloadLocations(h.URL, *m) //nolint
```
Comment (codefromthecrypt) [inline pkg/test/manifest/server.go:41]: this lint error will go away when we merge #214

File: test/e2e/getenvoy_run_test.go [hunk 4]
```diff
@@ -23,10 +23,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/mholt/archiver/v3"
```
Comment (codefromthecrypt) [inline test/e2e/getenvoy_run_test.go:26]: some time in the future, I want to make a separate go.mod file for the e2e tests, so that this dep doesn't clutter main

**Comment by codefromthecrypt** [comment]
ok I retouched this with an updated description

**Comment by codefromthecrypt** [comment]
I need to verify this manually prior to merge because GH actions doesn't actually download anything due to caching... `1.17.1/darwin is already downloaded`

**Comment by codefromthecrypt** [comment]
We have used xz compression except for the first release (which was gz), so I had to add a dependency for that.

File: pkg/binary/envoy/fetch_test.go [hunk 5]
```diff
@@ -17,26 +17,134 @@ package envoy
...
+		io.Copy(w, f)
+		w.WriteHeader(200)
+	}
+	t.Run("tar missing bin/envoy", func(t *testing.T) {
+		e := fetchEnvoy(envoyPath, server.URL, internal.GzipReader)
+		require.EqualError(t, e, fmt.Sprintf(`bin/envoy not found in %s`, server.URL))
+	})
+}
+
+func TestFetchEnvoy(t *testing.T) {
```
Comment (codefromthecrypt) [inline pkg/binary/envoy/fetch_test.go:93]: ended up having to add more tests as only now I realized we are downloading both xz and gz archives

**Comment by codefromthecrypt** [comment]
ok all good manually and also in CI

**Comment by codefromthecrypt** [comment]
Added a commit to split e2e from the top-level go.mod as it now makes a difference

File: test/e2e/go.mod [hunk 6]
```diff
@@ -0,0 +1,12 @@
+module github.com/tetratelabs/getenvoy/e2e
```
Comment (codefromthecrypt) [inline test/e2e/go.mod:1]: this is the new go.mod for e2e tests. As you can see archiver/v3 is here and we can also add anything we want/need without fear of accidentally bloating the binary. cc @musaprg @mathetake

File: test/e2e/main_test.go [hunk 7]
```diff
@@ -12,39 +12,42 @@
...
+	path, err := readGetEnvoyPath()
 	if err != nil {
 		fmt.Fprintf(os.Stderr, `failed to start e2e tests due to an invalid "getenvoy" binary: %v`, err)
 		os.Exit(1)
 	}
-	e2e.GetEnvoyPath = path
+	getEnvoyPath = path
+
+	// Now, lookup the version we are to use
+	referenceTxt := "../../pkg/reference.txt"
```
Comment (codefromthecrypt) [inline test/e2e/main_test.go:39]: This is reading the file instead of using module embedding intentionally as it is cleaner than mixing the top-level module into this.

File: test/e2e/getenvoy_test.go [hunk 8]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package e2e_test
+package e2e
```
Comment (codefromthecrypt) [inline test/e2e/getenvoy_test.go:15]: put in same package as utilities to avoid exporting everything

File: go.mod [hunk 9]
```diff
@@ -5,20 +5,15 @@ go 1.16
 
 require (
 	github.com/StackExchange/wmi v0.0.0-20210224194228-fe8f1750fd46 // indirect
-	github.com/andybalholm/brotli v1.0.2 // indirect
 	github.com/go-ole/go-ole v1.2.5 // indirect
-	github.com/klauspost/compress v1.12.2 // indirect
-	github.com/klauspost/pgzip v1.2.5 // indirect
 	github.com/mattn/go-runewidth v0.0.12 // indirect
-	github.com/mholt/archiver/v3 v3.5.0
```
Comment (codefromthecrypt) [inline go.mod:13]: now that e2e is split, there's no reference here anymore nor any indirect stuff in go.sum

**Comment by codefromthecrypt** [comment]
I rewrote the code and now it increases coverage albeit slightly (85.8% -> 85.9%)

**Comment by codefromthecrypt** [comment]
it doesn't seem like our implementation of progress bar likes to be tapped into a read stream (rather the write side). I'll look into it prior to merge.

**Comment by codefromthecrypt** [comment]
will pull out the general log and test fixture refactoring into a separate PR as it is unrelated to the archive dep, though needed for me to figure out what was going on.

**Comment by codefromthecrypt** [comment]
will rebase after merging https://github.com/tetratelabs/getenvoy/pull/220

File: pkg/internal/tar.go [hunk 10]
```diff
@@ -47,47 +48,94 @@ func NewDecompressor(src string, r io.Reader) (io.Reader, error) {
...
-	// No support for streaming https://github.com/mholt/archiver/pull/199
-	archive := filepath.Join(src, "archive.tar")
-	err := archiver.DefaultTar.Archive([]string{filepath.Join(src, root)}, archive)
-	if err != nil {
-		return err
-	}
-	defer os.Remove(archive) //nolint
+	basePath := filepath.Dir(root)
+	// Recurse through the path including all files and directories
+	return fs.WalkDir(src, root, func(path string, d os.DirEntry, err error) error {
```
Comment (codefromthecrypt) [inline pkg/internal/tar.go:104]: trivia: this is a go 1.16 form

**Comment by codefromthecrypt** [comment]
oddly coverage is exactly the same 85.6%

**Review by mathetake — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
#220 removed a lot of the bloat, so I updated the description to be more precise with what's going on. The last commit includes cleanups requested before and some new ones as I noticed progress bar behaves differently depending on the HTTP response (ex if content length is available or not)

**Comment by codefromthecrypt** [comment]
aggregate statement coverage is exactly the same before and after this: 86.9%

File: go.sum [hunk 11]
```diff
@@ -19,8 +19,6 @@ github.com/StackExchange/wmi v0.0.0-20210224194228-fe8f1750fd46 h1:5sXbqlSomvdjl
 github.com/StackExchange/wmi v0.0.0-20210224194228-fe8f1750fd46/go.mod h1:3eOhrUMpNV+6aFIbp5/iudMxNCF27Vw2OZgy4xEx0Fg=
 github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
 github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
-github.com/andybalholm/brotli v1.0.0 h1:7UCwP93aiSfvWpapti8g88vVVGp2qqtGyePsSuDafo4=
```
Comment (codefromthecrypt) [inline go.sum:22]: reduction in lines here simplify our understanding of dependencies that can be used

File: internal/tar/tar.go [hunk 12]
```diff
@@ -12,82 +12,153 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package tar avoids a large (~3MB) dependency on archiver/v3. These are special-cased to the needs of getenvoy.
```
Comment (codefromthecrypt) [inline internal/tar/tar.go:15]:
```suggestion
// Package tar avoids third-party dependencies (ex archiver/v3) and are special-cased to the needs of getenvoy.
```

File: internal/tar/tar.go [hunk 13]
```diff
@@ -12,82 +12,153 @@
...
-// Untar unarchives, stripping the base directory inside the "src" archive. Ex on "/foo/bar", "dst" will have "bar/**"
-func Untar(dst string, src io.Reader) error { // dst, src order like io.Copy
-	// No support for streaming https://github.com/mholt/archiver/pull/199
-	tmpDir, err := ioutil.TempDir("", "getenvoy-")
-	if err != nil {
-		return err
+// newDecompressor returns an "xz" or "gzip" decompression function based on bytes in the stream.
+func newDecompressor(r io.Reader) (io.ReadCloser, error) {
+	br := bufio.NewReader(r)
+	h, err := br.Peek(xz.HeaderLen)
```
Comment (codefromthecrypt) [inline internal/tar/tar.go:83]: @mathetake doing this now and backfilled some tests around corner cases

File: internal/tar/tar_test.go [hunk 14]
```diff
@@ -12,9 +12,11 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package tar_test
+package tar
```
Comment (codefromthecrypt) [inline internal/tar/tar_test.go:15]: to test un-exported functions

**Comment by mholt** [comment]
FYI archiver v4 (coming soon from https://github.com/mholt/archiver/pull/302) has an all-new stream oriented API. :+1:
